### PR TITLE
feat(ISV-5818): Attest the release-time SBOM to the registry.

### PR DESCRIPTION
**Pull request by BorekZnovustvoritel**

### CODE REVIEW

File: tests/test_oci.py [hunk 1]
```diff
@@ -596,3 +630,274 @@ async def mock_run_async_subprocess(
 
         with pytest.raises(SBOMError):
             await client.fetch_sbom(image)
+
+    @pytest.mark.asyncio
+    @patch("mobster.oci.cosign.run_async_subprocess")
+    @patch("mobster.oci.cosign.make_oci_auth_file")
+    async def test_attach_sbom_spdx(
```
Comment (jedinym) [inline tests/test_oci.py:637]: I know that this is a WIP but I'd like to chime in before you spend more time on this. 

I think these tests are very brittle, while not providing much benefit.

E.g. what happens when we decide to change the name/function signature of make_auth_file? This test breaks, even though all might be working fine. Or what if there's a edge case bug in cosign that makes it not do what we expect? 

This test does not verify that an sbom is attached and the later tests don't verify that sboms are attested/cleaned, just that _some_ code is run is _some_ specific sequence.

If we replaced these with integration tests, we would make them less brittle - we're only testing the edges of the system that are unlikely to change.

Plus we would get actual assurance that an SBOM was attached/attested/cleaned, because we would directly look at the registry data.

We could do that in integration/test_tekton.py or create integration/test_cosign.py.

The only problem is rekor, as we don't have that in our compose.yaml. In that case I would not test that specifically, since rekor is still not used anyway. We can postpone those tests to then.

Comment (BorekZnovustvoritel) [reply tests/test_oci.py]: I have tried to make the tests less hard-wired to the code, but I want to appease the coverage bot :grinning: So I have kept at least some unit-testing in-place

File: src/mobster/cmd/augment/__init__.py [hunk 2]
```diff
@@ -231,22 +238,49 @@ async def update_sbom(
...
             await write_sbom(sbom.doc, path)
+            sbom_format = sbom.format
 
             # run garbage collection manually to make sure the object is
             # cleaned up before we release the lock
             del sbom
             gc.collect()
 
-            LOGGER.info("Successfully enriched SBOM for image %s", image)
+            await config.cosign.clean(sbom_reference, "sbom")
```
[the remark this answers is not in the document]
Comment (jedinym) [reply src/mobster/cmd/augment/__init__.py]: Could you please summarize the offline discussion for me? :smile:

File: src/mobster/oci/cosign.py [hunk 3]
```diff
@@ -91,6 +146,27 @@ async def fetch_latest_provenance(self, image: Image) -> Provenance02:
...
+            sbom_format: The expected SBOM format. This function only cares if the type
+                is SPDX or CDX, specific version does not matter
+
+        Returns:
+            The initialized SBOM object
+        """
+        attestation_type = self.__get_cosign_attestation_type(sbom_format)
+        last_attestation = (
+            await self._perform_cosign_verify_attestation(image, attestation_type)
+        )[-1]
```
Comment (Allda) [inline src/mobster/oci/cosign.py:163]: I think this should verify if the attestation exists, otherwise, I think it can raise IndexError.

File: src/mobster/oci/cosign.py [hunk 4]
```diff
@@ -34,6 +39,38 @@ async def fetch_sbom(self, image: Image) -> SBOM:
         """
         raise NotImplementedError()
 
+    async def attest_sbom(
+        # pylint: disable=too-many-arguments,too-many-positional-arguments
```
Comment (jedinym) [inline src/mobster/oci/cosign.py:43]: We can avoid this warning by bundling the rekor args together into a dataclass. They can't be None independently anyway, just together.

File: src/mobster/tekton/component.py [hunk 5]
```diff
@@ -127,6 +191,49 @@ async def process_component_sboms(args: ProcessComponentArgs) -> None:
     artifact.write_result(args.result_dir)
 
 
+async def push_sbom_to_registry(
```
Comment (jedinym) [inline src/mobster/tekton/component.py:194]: Nitpick: I think we should go with `attest_sbom_to_registry` to make a bit clearer.

File: src/mobster/tekton/component.py [hunk 6]
```diff
@@ -127,6 +191,49 @@ async def process_component_sboms(args: ProcessComponentArgs) -> None:
...
+                    "Successfully attested image %s.", sbom_ref_detail.reference
+                )
+            except SBOMError:
+                LOGGER.exception("Could not attest SBOM because of a cosign error.")
+                return False
+            except Exception:  # pylint: disable=broad-exception-caught
+                LOGGER.exception("Could not attest SBOM because of an unknown error.")
+                return False
+    else:
+        LOGGER.warning("Could not attest image %s!", sbom_ref_detail.reference)
```
Comment (jedinym) [inline src/mobster/tekton/component.py:233]: This doesn't have a clear indicator of _why_ we cannot attest images, and it then returns True, which doesn't seem right. In the calling function we assign the signing key to the cosign client, so this branch shouldn't even ever run.

From the POV of the release pipeline, the attestations are mandatory so we shouldn't make it possible to not attest SBOMs.

File: tests/test_oci.py [hunk 7]
```diff
@@ -512,6 +515,38 @@ async def mock_run_async_subprocess(
...
+            "predicateType": "https://in-toto.io/Statement/v0.1",
+            "predicate": sbom_doc,
+        }
+        expected_attestation = {
+            "payload": base64.b64encode(json.dumps(expected_payload).encode()).decode()
+        }
+        attestation_raw = json.dumps(expected_attestation).encode()
+
+        async def mock_perform_cosign_verify_attestation(*_: Any) -> list[bytes]:
+            return [attestation_raw]
```
Comment (jedinym) [inline tests/test_oci.py:536]: Nitpick: This is not a mock, but a stub.
https://www.martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs

```suggestion
        async def fake_perform_cosign_verify_attestation(*_: Any) -> list[bytes]:
            return [attestation_raw]
```

**Review by jedinym — CHANGES_REQUESTED** [review]
Nice job on the tests, they're way better.

Let me know how the integration with release pipeline goes.

File: src/mobster/tekton/component.py [hunk 8]
```diff
@@ -78,23 +123,38 @@ def augment_component_sboms(
...
+    )
+    result_details = await update_sboms(config, snapshot)
+    if not all(result_details):
+        raise SBOMError("Could not generate all SBOMs!")
+    push_tasks = [
+        push_sbom_to_registry(
+            result_detail, cosign_client, semaphore, rekor_url, rekor_key
+        )
+        for result_detail in result_details
+        if result_detail
```
[the remark this answers is not in the document]
Comment (BorekZnovustvoritel) [reply src/mobster/tekton/component.py:153]: Yes :sweat_smile: if you prefer a different solution, I am very much willing to adapt it

[the remark this answers is not in the document]
Comment (jedinym) [reply src/mobster/tekton/component.py:153]: Huh, weird that it doesn't pick up on the code above. Since it is just for mypy and not any logic we need, I think I'd prefer a mypy ignore comment.

File: src/mobster/tekton/component.py [hunk 9]
```diff
@@ -78,23 +123,38 @@ def augment_component_sboms(
...
-
-    subprocess.run(cmd, check=True)
+    semaphore = asyncio.Semaphore(concurrency)
+    snapshot = await make_snapshot(snapshot_spec, None, semaphore)
+    cosign_client = CosignClient(
+        signing_key=sign_key, signing_key_password=sign_password
+    )
+    config = AugmentConfig(
+        cosign=cosign_client,
+        verify=False,
```
[the remark this answers is not in the document]
Comment (BorekZnovustvoritel) [reply src/mobster/tekton/component.py:140]: I have just one concern: Previously the validation wasn't required, but now if I change it to require the validation, an error is raised in case of a validation failure (e.g. the Provenance is missing). Should I catch and log these validation errors (perform a soft validation) or should I really make the whole script fail?

[the remark this answers is not in the document]
Comment (BorekZnovustvoritel) [reply src/mobster/tekton/component.py:140]: Agreed, will do.

File: src/mobster/cmd/augment/__init__.py [hunk 10]
```diff
@@ -28,6 +28,24 @@
...
+    The SBOMs are not stored in-memory but in
+    the filesystem to ease the load on memory.
+
+    Attributes:
+        reference: Reference to the image this attestation belongs to
+        sbom_format: The format of the SBOM
+        path: Path to the local SBOM file
+    """
+
+    reference: str
```
[the remark this answers is not in the document]
Comment (BorekZnovustvoritel) [reply src/mobster/cmd/augment/__init__.py:45]: It's a good suggestion, I will take a look at this after I am certain that the integration tests pass. I haven't addressed it as of now yet though.

File: src/mobster/tekton/component.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (BorekZnovustvoritel) [reply src/mobster/tekton/component.py]: I have implemented this and now the integration test for a big release won't pass because of OOM-killed error. I guess I am in for a great time of concurrency-throttling for cosign. Ruh-roh

File: src/mobster/release.py [hunk 11]
```diff
@@ -161,6 +167,7 @@ class ComponentModel(pdc.BaseModel):
 
     name: str
     image_reference: str = pdc.Field(alias="containerImage")
+    repository: str | None = pdc.Field(default=None)
```
Comment (jedinym) [inline src/mobster/release.py:170]: This is a required field (based on apply-mapping), we should not allow it to be missing.

**Review by jedinym — CHANGES_REQUESTED** [review]
Overall looks good, thank you for the big amount of time you spent on researching how to implement this!

File: tests/integration/oci_client.py [hunk 12]
```diff
@@ -226,6 +234,32 @@ async def attach_sbom(
...
+                                        + sbom_blob_digest,
+                                    },
+                                    {"name": "IMAGE_DIGEST", "value": image.digest},
+                                ],
+                            }
+                        ]
+                    },
+                }
+            )
+            await cosign_client.attest_provenance(provenance, image.reference)
```
[the remark this answers is not in the document]
Comment (jedinym) [reply tests/integration/oci_client.py:262]: We also use the oci_client in cases where we don't need to use the provenance (in the upcoming contextual parent integration tests. Really my main problem is that a client that's supposed to do simple registry stuff gets passed another client to do something closely related, but not really. Yes, it's just for integration tests, but it still shouldn't become a god object :smile: I'd prefer if we split it up somewhere else.

[the remark this answers is not in the document]
Comment (BorekZnovustvoritel) [reply tests/integration/oci_client.py]: You have convinced me. Updated :saluting_face:

**Review by jedinym — APPROVED** [review]
